### ISSUE TITLE
Inconsistent syntax

### ISSUE THREAD

**Issue by Udi-Fogiel** [reutenauer/polyglossia#610]
Currently the source contains several methods for various tasks, e.g. for defining commands there are TeX primitives (`\def`, `\let` etc.), LaTeX2e commands (e.g. `\newcommand`), LaTeX3 commands (e.g. `\NewDocumentCommand`), expl3 functions (e.g. `\cs_new:n`) and commands from the `makecmds` package (e.g. `\provideenvironment`).  I think the source would be much more readable if it was more consistent, with well structured syntax rules. 

There are several aspects to consider. The first, is what methods to use. For example, the following essentially do the same
 https://github.com/reutenauer/polyglossia/blob/3737629d973c1b197b9c0342df6877555c4a3954/tex/polyglossia.sty#L2648 
https://github.com/reutenauer/polyglossia/blob/3737629d973c1b197b9c0342df6877555c4a3954/tex/polyglossia.sty#L1436-L1437
but in a first glance they look rather different. 

Another aspect is macros (or functions) names. The source have two prefixes for internal commands, `xpg` and `polyglossia`. I'm not sure what is the difference, nor what is the meaning of `xpg`.  It does not seem like the naming conventions specified in `interface3` (section 1.1) are respected (using `@` for function names, using `\DeclareDocumentCommand` for private functions) and scratch variable syntax can be improved.

IMO the best approach would be to rewrite everything in `expl3` and LaTeX3 syntax, and to use `l3keys2e` only (not xkeyval). This is because most of the code is already dependent on it, and it offers more functionality than the other options.

It probably should be done is several stages, while testing after each stage.

**Comment by jspitz** [comment]
I have also started to transform things to LaTeX3 or modern LaTeX syntax. But it is quite some work and needs to be tested carefully. We also need to consider with some internal macros that other packages might patch them (that's why I refrained from unifying the `xpg` and `polyglossia` namespace until now; `xpg` is the old one, I think it means "e**x**ternal package **p**oly**g**lossia" or maybe "**X**eteX **p**oly**g**lossia" given that this was the original scope or something along these lines. Of course `polyglossia` is more transparent, but `xpg` is more economic and quite a neat acronym, so I tend to go for that at least for internal pseudo-namespace)

Another thing on my agenda is getting rid of `xkeyval` and using `l3keys[2e]` instead, but we can only do that once for all gloss files.

**Comment by Udi-Fogiel** [comment]
>We also need to consider with some internal macros that other packages might patch them (that's why I refrained from unifying the `xpg` and `polyglossia` namespace until now;

Hmm I'm not sure what to do about that other than make a public statement. But this can wait, we can start with dealing with the private internal functinos, and removing unnecessary dependencies (such as `expl3` package, or [luatexbase](https://tex.stackexchange.com/questions/681353/polyglossia-loads-luatexbase)).

**Comment by Udi-Fogiel** [comment]
> Another thing on my agenda is getting rid of `xkeyval` and using `l3keys[2e]` instead, but we can only do that once for all gloss files.

Does this mean that gloss files will eventually use expl3 syntax?

**Comment by Udi-Fogiel** [comment]
> I would rather avoid a general rewrite of the gloss files. They need to be accessible to their respective maintainers. 

sure, I agree.

>I think the LaTeX2e l3keys interface will do for most, even though it doesn't provide for choice keys yet.

Is there one? how can I define a new key in "normal"catcode regime?

**Comment by jspitz** [comment]
> Is there one? how can I define a new key in "normal"catcode regime?

`\DeclareKeys[⟨family⟩]{⟨declarations⟩}`

(and `\ProcessKeyOptions`, `\SetKeys` etc.)

See `clsguide`, sec. 4.4

This has been introduced in LaTeX release 2022-06-01

**Comment by Udi-Fogiel** [comment]
Oh, this is great! I did not know about this.

**Comment by jspitz** [comment]
Yes, it's pretty neat. I hope they'll extend the syntax,

**Comment by Udi-Fogiel** [comment]
what is the reason for patching `\XKV@define@default`? will we still need something similar?

**Comment by jspitz** [comment]
> what is the reason for patching `\XKV@define@default`? 

See #364 and f71fc8166f12df7c8f

> will we still need something similar?

We need a way to access the default value associated with a key (in case no value is given). On a quick glance I haven't found something useful in `l3keys`.

**Comment by Udi-Fogiel** [comment]
I don't think there is a notion of 'default value' in `\DeclareKeys`. Maybe we can define our own parser. see section 27.10 of interface3.

**Comment by Udi-Fogiel** [comment]
yes, but I thought we are trying to not enforce gloss files to use L3 interface.

**Comment by jspitz** [comment]
I have never tried actually, but `clsguide` states that
```
A small number of ‘basic’ ⟨properties⟩ are described below.
The full range of properties, provided by l3keys, can also be used for more
powerful processing. See interface3 for the full details.
```
I am not sure whether this means that the `l3keys` properties can be passed as is to  `\DeclareKeys`, but it looks so (we just need to try).

**Comment by Udi-Fogiel** [comment]
works!

**Comment by Udi-Fogiel** [comment]
works for choice keys as well

**Comment by Udi-Fogiel** [comment]
> We need a way to access the default value associated with a key (in case no value is given). On a quick glance I haven't found something useful in `l3keys`.

Here is an example to explain how we can do that
```TeX
\documentclass{article}

\DeclareKeys[foo]{bar.code:n = #1, bar.default:n = baz}

\begin{document}
\csname\detokenize{key default > foo/bar}\endcsname

\end{document}
```

**Comment by Udi-Fogiel** [comment]
@jspitz I'm working on a transition to l3keys, but I don't understand something. Why the difference in this example:
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{french}
\begin{document}
\selectlanguage[variant=canadian]{french}

\babelname

\selectlanguage{french}

\babelname

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

a?

\selectlanguage[autospacing=false]{french}
a?

\selectlanguage{french}
a?

\end{document}
```
i.e. why the variant is restored but the autospacing is not? This also implies that if I will use an alias with `autospacing=false` it will produce unexpected result
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{french}
\setlanguagealias[autospacing=false]{french}{foo}
\begin{document}

a?

\selectlanguage{foo}
a?

\selectlanguage{french}
a?

\end{document}
```

To me, it looks like a bug, but I'm not sure. I always though that the options selected at language loading time are the default ones for this language.

**Comment by jspitz** [comment]
I am not sure. With the language environments and inline commands (`\textfrench`), options are local. With `\selectlanguage` as well as \foreignlanguage and maybe others current options are inherited.

I am not sure whether this is explicit design, and I am also not sure whether I like it or not. But it is established behavior, and we will interfere with people's documents if we change this now (the solution would probably be as easy as clearing the option list and resetting defaults at the beginning of `\xpg_alias_add_to_option_i:nn`).

**Comment by Udi-Fogiel** [comment]
The documentation says otherwise:
>The command \foreinlanguage{⟨lang⟩}{⟨…⟩} and the
otherlanguage* environment are identical with the use of the \text⟨lang⟩ or
\textlang command, with the one notable exception that they do not translate
the date with \today.

**Comment by jspitz** [comment]
Yes, but it behaved differently.

**Comment by Udi-Fogiel** [comment]
Then to me it is clearly a bug :) In any case, can you point to where variants are stored and recovered?, as they are clearly an exception here.

**Comment by jspitz** [comment]
`\__xpg_set_language_options:nnn`

**Comment by Udi-Fogiel** [comment]
is it stored in `\angvariant`? this is the only difference I see from other keys.

**Comment by Udi-Fogiel** [comment]
Ok, it probably isn't, there is just a bug in `\babelname`...
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage[variant=canadian]{french}
\begin{document}

\babelname

\selectlanguage[variant=acadian]{french}

\babelname

\selectlanguage{french}
\languagevariant


\end{document}
```

**Comment by Udi-Fogiel** [comment]
Or maybe it is restored?
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage[variant=canadian]{french}
\begin{document}

\languagevariant

\babelname

\BCPdata{tag}

\selectlanguage[variant=acadian]{french}
\languagevariant

\babelname

\BCPdata{tag}

\selectlanguage{french}
\languagevariant

\babelname

\BCPdata{tag}


\end{document}
```

but then `\languagevariant` is bugged.

**Comment by jspitz** [comment]
Both `\languagevariant` and `\babelname` are not correctly set on `\selectlanguage`

**Comment by Udi-Fogiel** [comment]
and `\BCPdata`?

**Comment by jspitz** [comment]
> and `\BCPdata`?

also :-(

**Comment by Udi-Fogiel** [comment]
I think we should consider changing the behavior of `\foreignlanguage` and `\selectlanguage`...

**Comment by jspitz** [comment]
This is a bug in the repo which isn't in the last release. I'll check.

**Comment by Udi-Fogiel** [comment]
And it looks like `\foreignlanguage` changes the parameters locally. 
```
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{hebrew}
\setmainfont{David CLM}
\begin{document}
\textdir TLT
\foreignlanguage[numerals=hebrew]{hebrew}{\localnumeral{12}}

\foreignlanguage{hebrew}{\localnumeral{12}}
\end{document}
```

![image](https://github.com/reutenauer/polyglossia/assets/110533647/ba6b9563-8319-4b3c-9923-e31922533915)






> This is a bug in the repo which isn't in the last release. I'll check.

Oh, I should have mentioned, I tested everything with `v.165`, so I'm not sure what is the correct result anymore. I'll update my prevoius comments with pictures.

**Comment by jspitz** [comment]
git bisect points out 55de651eeba as the commit that introduced the bug

**Comment by Udi-Fogiel** [comment]
Again, I'm not sure how you deduced that because I don't have an example with expected result, [this picture](https://private-user-images.githubusercontent.com/110533647/292630444-3bee01cc-85cf-4144-9be6-a98f396c140c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzNDE1OTYsIm5iZiI6MTcwMzM0MTI5NiwicGF0aCI6Ii8xMTA1MzM2NDcvMjkyNjMwNDQ0LTNiZWUwMWNjLTg1Y2YtNDE0NC05YmU2LWE5OGYzOTZjMTQwYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyM1QxNDIxMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NTUzOWVmNGFiMjA0NTVlZDI3ZmI3NGQ2ZmJkNzhmNzNmZjI4Zjk5ZTc2OWZlMmUzMjgyYTViOWU0MTU2MzRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nwzHOQ4HD6PnWu-Vsqa4glqZtN0ouRGtHS7Dj07olV8) is obviously wrong and produced with `v1.65`... If you have a good example, Ill be hapy to fix that.

**Comment by Udi-Fogiel** [comment]
And if the `\babelname` in the linked picture is correct, how come it gets restored, and all other information related to the variant, but other keys are inherited?

**Comment by jspitz** [comment]
Take this:
```tex
% !TeX TS-program = xelatex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage[variant=canadian]{french}
\setotherlanguage{german}
\begin{document}
	
\section{Canadian French}
	Variant: \languagevariant
	
	Babelname: \babelname
	
	BCP-47: \BCPdata{tag}

\section{Acadian French}
	\selectlanguage[variant=acadian]{french}
	Variant: \languagevariant
	
	Babelname: \babelname
	
	BCP-47: \BCPdata{tag}

\section{No Option French (= Canadian)}
	\selectlanguage{french}
	Variant: \languagevariant
	
	Babelname: \babelname
	
	BCP-47: \BCPdata{tag}

\section{Austrian German}
	\selectlanguage[variant=austrian]{german}
	Variant: \languagevariant
	
	Babelname: \babelname
	
	BCP-47: \BCPdata{tag}
	
\end{document}
```

With 1.66:
![grafik](https://github.com/reutenauer/polyglossia/assets/4538783/123e6641-e3cf-40c7-80bc-d7ec5d23ff7b)

With current master:
![grafik](https://github.com/reutenauer/polyglossia/assets/4538783/8843be66-9d73-4ea0-82ae-949e4dbdcab4)

As you can see, the babelname and BCP47 value is wring, variant is correct.

**Comment by jspitz** [comment]
More correctly, variant is wrong in both cases for the option-less switch (sec. 3)

**Comment by jspitz** [comment]
> git bisect points out [55de651](https://github.com/reutenauer/polyglossia/commit/55de651eeba716611736fa6cc2c17607c93772b6) as the commit that introduced the bug

I have fixed this at fe72d6405db416f52f

**Comment by Udi-Fogiel** [comment]
> I have fixed this at [fe72d64](https://github.com/reutenauer/polyglossia/commit/fe72d6405db416f52f00e87f97b8c2d9798b603d)

was just about to push the same. Thanks!

**Comment by Udi-Fogiel** [comment]
> And if the `\babelname` in the linked picture is correct, how come it gets restored, and all other information related to the variant, but other keys are inherited?

It still does not explain why all the information regarding the variant key is restored but all other keys are not.

**Comment by Udi-Fogiel** [comment]
> And it looks like `\foreignlanguage` changes the parameters locally.
> 
> ```
> \documentclass{article}
> \usepackage{polyglossia}
> \setmainlanguage{hebrew}
> \setmainfont{David CLM}
> \begin{document}
> \foreignlanguage[numerals=hebrew]{hebrew}{\localnumeral{12}}
> 
> \foreignlanguage{hebrew}{\localnumeral{12}}
> \end{document}
> ```

Never mind, now I understand what you meant.

**Comment by Udi-Fogiel** [comment]
> Take this:
> 
> ```
> % !TeX TS-program = xelatex
> \documentclass{article}
> \usepackage{polyglossia}
> \setmainlanguage[variant=canadian]{french}
> \setotherlanguage{german}
> \begin{document}
> 	
> \section{Canadian French}
> 	Variant: \languagevariant
> 	
> 	Babelname: \babelname
> 	
> 	BCP-47: \BCPdata{tag}
> 
> \section{Acadian French}
> 	\selectlanguage[variant=acadian]{french}
> 	Variant: \languagevariant
> 	
> 	Babelname: \babelname
> 	
> 	BCP-47: \BCPdata{tag}
> 
> \section{No Option French (= Canadian)}
> 	\selectlanguage{french}
> 	Variant: \languagevariant
> 	
> 	Babelname: \babelname
> 	
> 	BCP-47: \BCPdata{tag}
> 
> \section{Austrian German}
> 	\selectlanguage[variant=austrian]{german}
> 	Variant: \languagevariant
> 	
> 	Babelname: \babelname
> 	
> 	BCP-47: \BCPdata{tag}
> 	
> \end{document}
> ```

Can you explain why in `\section{No Option French (= Canadian)}` the result of `\babelname` is canadien? From
>With `\selectlanguage` as well as `\foreignlanguage` and maybe others current options are inherited.

I expected the variant to be inherited as well...

**Comment by jspitz** [comment]
The variant inheritance is fixed at 54000b1e0b40c

**Comment by Udi-Fogiel** [comment]
I'm taking about the key, not the command `\languagevariant`. Let's go back to my first example
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{french}
\begin{document}
\selectlanguage[variant=canadian]{french}

\babelname

\selectlanguage{french}

\babelname

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

a?

\selectlanguage[autospacing=false]{french}
a?

\selectlanguage{french}
a?

\end{document}
```

![image](https://github.com/reutenauer/polyglossia/assets/110533647/55a6ef95-0564-465d-aa89-4ac620787855)

Do you see that `autospacing`'s value is inherited, but the `variant`'s is not?

**Comment by jspitz** [comment]
But you agree that what I fixed was a bug as well?
Yes, I see that `autospacing` should be reset to default here, but isn't.

**Comment by jspitz** [comment]
I suppose we need to store the keys in ` \__xpg_load_lang_definition:nn` in restore them in case a language is switched to without options.

**Comment by jspitz** [comment]
We do save the current options of a language in `xpg@current@options@<lang>`. If we'd do the same for the initial setting of a language, we could restore that.

**Comment by jspitz** [comment]
(see for instance the otherlanguage environment, where we restore the pre-switch options after the environment has been closed)

**Comment by Udi-Fogiel** [comment]
> But you agree that what I fixed was a bug as well?

Of course!

 >Yes, I see that `autospacing` should be reset to default here, but isn't.

Ok, that has answered my initial question. so what did you mean by the following?

> With \selectlanguage as well as \foreignlanguage and maybe others current options are inherited.

that these inherit the initial keys?

**Comment by Udi-Fogiel** [comment]
> that these inherit the initial keys?

If this is the case, I did not understand correctly, sorry... I do agree that it is ok to keep it this way, and not considering as a bug. 


Now that everything is clear I will resume the work on the transition to to `l3keys`, I'll probably just fix the bug while doing it.

**Comment by Udi-Fogiel** [comment]
> (see for instance the otherlanguage environment, where we restore the pre-switch options after the environment has been closed)

I'm wondering why this is needed, as `\selectlanguage` should handle the restoring of the settings with `\aftergroup`.

**Comment by Udi-Fogiel** [comment]
Ok, this is because the only thing that happens at the end of the group with `\selectlanguage` is basically this:
https://github.com/reutenauer/polyglossia/blob/fb7d43c38bc713bae410b7076fc1afb50af5853f/tex/polyglossia.sty#L1904-L1907
And it seems like something is not restoring correctly there. It also means we write to the aux file twice sometimes...

**Comment by Udi-Fogiel** [comment]
Actually, it seems like nested `\selectlanguae` are ok, 
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage[numerals=arabic]{hebrew}
\setmainfont{David CLM}
\protected\def\foo#1{}
\begin{document}
\makeatletter
\def\bar#1{\write\@auxout{\foo{#1}}}
\bar{arabic}
{
\selectlanguage[numerals=hebrew]{hebrew}
\bar{hebrew}
{
\selectlanguage[numerals=arabic]{hebrew}
\bar{arabic}
}\bar{hebrew}}
\bar{arabic}
a
\end{document}
```
Here is the aux
```
\relax 
\ProvideDocumentCommand \selectlanguage {sO{}m}{}
\@writefile{toc}{\selectlanguage *[numerals=arabic]{hebrew}}
\foo {arabic}
\@writefile{toc}{\selectlanguage *[numerals=hebrew]{hebrew}}
\foo {hebrew}
\@writefile{toc}{\selectlanguage *[numerals=arabic]{hebrew}}
\foo {arabic}
\@writefile{toc}{\selectlanguage *[numerals=hebrew]{hebrew}}
\foo {hebrew}
\@writefile{toc}{\selectlanguage *[numerals=arabic]{hebrew}}
\foo {arabic}
\gdef \@abspage@last{1}
```

maybe the fix at 1e6f728 related to nested languages is not really needed, and the source of it is the bug with the default values, and it fixed it seems to fix it because it consider default values?

**Comment by jspitz** [comment]
Please go ahead with `l3keys` transition. We can address bugs with key restoration later. If the language stacking turns out to be not needed, and grouping works, the better (I didn't manage when I tried back then).

To make my point clear: I think that any `\selectlanguage` should start off from the initial key settings (as done in `\setmainlanguage` or `\setotherlanguage`). We just need to cache those, as we do with variable and babel name. But this doesn't need to be fixed now.

**Comment by jspitz** [comment]
I'll go on halt with my own rewrite for the time being in order to not step on your toes

**Comment by jspitz** [comment]
Apart from keyval, we are now basically done. We have used the LyX testcases to check nothing breaks at least in the construct the LyX test files use and the `polyglossia` languages LyX supports. We have not checked for consistency and issues in the output that do not error or warn.

The polyglossia style file and the glosses that are supported by LyX do not use `etoolbox` any longer. We'd need to check the remaining newer languages and then we can remove the call of the package.

Of course, any polishing and testing is highly welcome.

**Comment by Udi-Fogiel** [comment]
> > The polyglossia style file and the glosses that are supported by LyX do not use `etoolbox` any longer. We'd need to check the remaining newer languages and then we can remove the call of the package.

That's great! thank you for working on that, and sorry for disappearing in the last several months. 

> Update on that: a number of gloss files at least employ `\ifcsdef` and `\ifcsunded`. For these, we could provide LaTeX2e aliases for the respective l3 checks.

Well. the LaTeX kernel provides `\@ifundefined` which is equivalent to `\ifcsundef`, so do we really need to provide another equivalent? Note that both  `\ifcsdef` and `\ifcsundef` returns true if the command the equivalent to `\relax`, I don't know if it matters in the gloss files, from a quick glance it isn't, but replacing `\ifcsdef` with `\@ifundefined` should be done carefully.

**Comment by Udi-Fogiel** [comment]
> Apart from keyval, we are now basically done. 

Hopefully I will have time to finish the work on keyval in the next few days.

**Comment by jspitz** [comment]
> Well. the LaTeX kernel provides `\@ifundefined` which is equivalent to `\ifcsundef`, so do we really need to provide another equivalent? 

I remember it didn't work good particularly in the very nested cases. And I would like to avoid too much `\if`...`\fi` and `\csname`...`\endcsname`. The code is already hard to parse without that.

**Comment by Udi-Fogiel** [comment]
> > Well. the LaTeX kernel provides `\@ifundefined` which is equivalent to `\ifcsundef`, so do we really need to provide another equivalent?
> 
> I remember it didn't work good particularly in the very nested cases. And I would like to avoid too much `\if`...`\fi` and `\csname`...`\endcsname`. The code is already hard to parse without that.


As I mentioned, `\@ifundefined` is equivalent to `\ifcsundef` , i.e. it takes arguments, doesn't use `\if ... \fi`, it is ok for nesting.

When etoolbox as written, LaTeX did not assume yet that the user has `e-TeX`, hence the definition of `\@ifundefined` was different back then.

**Comment by Udi-Fogiel** [comment]
Some glosses also have `\providebool`, `ifbool` etc.

**Comment by jspitz** [comment]
> Some glosses also have `\providebool`, `ifbool` etc.

The `\providebool` cases have been addressed. But there is more complex stuff (such as `\patchcmd`) also in the glosses that tinker with the headers. I wonder whether we should introduce `\RequireEToolBox` and let those glosses use that.

**Comment by Udi-Fogiel** [comment]
That could potentially be a problem if someone is loading a language after the beginning of the document (#614), there the glosses will not be able to load any package.

**Comment by jspitz** [comment]
Right, but we'd have the same problem with `\RequireBidi`. So probably the answer to your question in #614 should be: no, languages must not be loaded in the document body.

**Comment by Udi-Fogiel** [comment]
Yes, but `\RequireBidi` always had this behavior, while introducing `\RequireEToolBox` can poteontially affect existing documents. Don't get me wrong, I'm in favor of limiting the loading time to the preamble, it will make our life much easier, and I don't see any disadvantage in that, apart from the possible errors that users can get (and can easily fix).

**Comment by jspitz** [comment]
Let's see. Getting rid of etoolbox is not the most important thing. I'd just be nice to have a package less in dependencies.

**Comment by jspitz** [comment]
I have removed the easy `etoolbox` remainders at 787195f98a766fbe5.

We only have `etoolbox` commands now in the glosses that patch the headings and therefore employ `\patchcmd`, namely French, Hungarian, Korean, and Uyghur.

**Comment by Udi-Fogiel** [comment]
I think I also saw `\ifstreq` in the Chinese gloss keys definition, I'm currently working on removing `xkeyval`  so I removed that on my local machine, but there might be more instances we missed. 

I guess the easiest way to check that would be to remove `etoolbox` from the core package and see which tests fail (note that currently the tests don't check errors at loading time, I changed that on my local machine as well).

I have 30 glosses more to remove `xkeyval` from, @jspitz do you that I'll push what I did up until now or only when I'll finish?

**Comment by Udi-Fogiel** [comment]
Actually, would you mind reverting the last two commits, there are conflicts I'm having trouble with.

**Comment by Udi-Fogiel** [comment]
> Actually, would you mind reverting the last two commits, there are conflicts I'm having trouble with.

never mind, I managed to resolve them.

**Comment by Udi-Fogiel** [comment]
The following are the remaining glosses that need xkeyval to be removed from
```
tex/gloss-latin.ldf
tex/gloss-german.ldf
tex/gloss-arabic.ldf
tex/gloss-persian.ldf
tex/gloss-armenian.ldf
tex/gloss-korean.ldf
tex/gloss-punjabi.ldf
tex/gloss-welsh.ldf
tex/gloss-odia.ldf
tex/gloss-catalan.ldf
tex/gloss-hebrew.ldf
tex/gloss-malay.ldf
tex/gloss-serbian.ldf
tex/gloss-uyghur.ldf
tex/gloss-english.ldf
tex/gloss-norwegian.ldf
tex/gloss-portuguese.ldf
tex/gloss-italian.ldf
tex/gloss-lao.ldf
tex/gloss-french.ldf
tex/gloss-slovenian.ldf
tex/gloss-syriac.ldf
tex/gloss-ukrainian.ldf
tex/gloss-gaelic.ldf
tex/gloss-thai.ldf
tex/gloss-mongolian.ldf
tex/gloss-dutch.ldf
tex/gloss-russian.ldf
tex/gloss-sanskrit.ldf
tex/gloss-georgian.ldf
```

**Comment by jspitz** [comment]
Sorry for the conflict, I won't commit anything now until you are done. I'd prefer to have all in at once to have a functioning package. But you could set up a branch and commit to those subsequently, and merge the branch with master once everything is ready and tested.

**Comment by Udi-Fogiel** [comment]
I've removed `xkeyval` from all the glosses. I've noticed that urdu, khmer, kannada, occitan, and piedmontese does not contain `\InitializeGlossOptions`, is this intentional?

**Comment by jspitz** [comment]
> I've removed `xkeyval` from all the glosses. 

Excellent, thanks.

> I've noticed that urdu, khmer, kannada, occitan, and piedmontese does not contain `\InitializeGlossOptions`, is this intentional?

This should probably be added.

**Comment by jspitz** [comment]
> There are instances of `\docvlist` in some glosses as well

I have removed these with 24132c6fb549bd

**Comment by Udi-Fogiel** [comment]
> I have removed these with [24132c6](https://github.com/reutenauer/polyglossia/commit/24132c6fb549bde8369fbd53335c536c88823730)

The malay gloss test fails after this commit, the warnings in the log seems strange to me, can you check that?

**Comment by jspitz** [comment]
Still don't know how to run the test suite, but a simple malay file compiles for me wihout problems.

**Comment by Udi-Fogiel** [comment]
It compiles, but there there are many warning messages. There are also more differences, I get different hyphenation in the luatex output of `test-gloss-dsb`.

To run the tests you can use `l3build check` to run all of them, if there are any fails, there should be .diff files in `build/test` and `build/test-configfiles`

**Comment by Udi-Fogiel** [comment]
To check a specific file from the auto generated file you can use e.g. `l3build check -c configfiles/config-autogen test-gloss-dsb` or if you want to test only with luatex then `l3build check -e luatex -c configfiles/config-autogen test-gloss-dsb`

**Comment by Udi-Fogiel** [comment]
Here is a simple test:
```tex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{dsb}
\begin{document}
\day=6
\month=8
\year=2012
\hsize=1cm
\today.
\end{document}
```
The output before 24132c6 is
![image](https://github.com/reutenauer/polyglossia/assets/110533647/c67c116a-4e53-41fb-b2b3-b11c63ec8f97)

and after the commit is
![image](https://github.com/reutenauer/polyglossia/assets/110533647/e680e037-d49b-4dd5-9202-718c94e55743)

**Comment by Udi-Fogiel** [comment]
I'll probably won't be available in the following several hours, if you won't find the problem I'll look deeper later.

**Comment by jspitz** [comment]
Should be fixed with e0518ea5b1c6ff. Thinko.

**Comment by Udi-Fogiel** [comment]
Looks good now, thanks. 

Just a question, I've noticed you added information to the log, is this desirable? (just checking before I update the tests for that)

And with `test-gloss-dsb` I'm seeing both the lines 

```
Package polyglossia Info: Using hyphenation pattern 'uppersorbian' for usorbian
```

```
Package polyglossia Info: Using hyphenation pattern 'usorbian' for lsorbian
```
which to me looks confusing, but I'm not sure if it is appropriate or not. 

The log also contains the following (not related to recent commits)
```
Package polyglossia Info: Option: sorbian, variant=upper
```
```
Package polyglossia Info: Option: sorbian, variant=lower
```
and the latter appears twice, I'm wondering why we need both lines, and why the second one appears twice. Maybe we should clean the information we write to the log.

**Comment by jspitz** [comment]
I thought it would be information that might be useful, particularly with cases that fall back to variant patterns (such as english -> newzealand). But I will remove it. On the other hand, I think the tests a quite a bit too pedantic here.

**Comment by Udi-Fogiel** [comment]
> I thought it would be information that might be useful, particularly with cases that fall back to variant patterns (such as english -> newzealand).

Then I'll update the tests. Just to be clear, the fact that I'm seeing this message twice with different information in the log is expected?

 >But I will remove it. On the other hand, I think the tests a quite a bit too pedantic here.

You don't need to remove that I just wanted to make sure that the current output is the expected one.

I agree that the tests are quite pedantic, but I'm not sure how to filter out such changes without also ignoring regressions like the one we had saw yesterday...

**Comment by jspitz** [comment]
> Just to be clear, the fact that I'm seeing this message twice with different information in the log is expected?

No.

> You don't need to remove that I just wanted to make sure that the current output is the expected one.

It's fine. I thought about removing it anyway.

**Comment by jspitz** [comment]
> > Just to be clear, the fact that I'm seeing this message twice with different information in the log is expected?
> 
> No.

Well, in a way, it is. I think the first message is issued with
```tex
\InitializeGlossOptions{sorbian}{variant=upper,olddate=false}
```
when the gloss is loaded. The others I would need to check. This probably is another argument why we should remove these info messages from the gloss and have it centralized in some function (`\SetGlossOptions`?).

**Comment by Udi-Fogiel** [comment]
@jspitz I find the code for for tracing `\babelname` prety complex, is there a reason why 
```tex
\DeclareExpandableDocumentCommand \babelname { }
  {
    \prop_item:Nn  \g_xpg_langsetup_prop { \languagename / babelname }
  }
```
won't suffice?

**Comment by Udi-Fogiel** [comment]
And what is the use of `\<lang>@gbabelname`?

**Comment by Udi-Fogiel** [comment]
The source of the problem described in #334 is the fact that we store things in `\g_xpg_langsetup_prop` globally. Since the variant key do these assignment there is a global change when TeX reads the .toc file, it changes the babelname entry globally, since the last line in the .toc file is `\selectlanguage *[variant=american]{english}`, `\babelname` is wrongly set. 

Is there a reason we do all these global assignments for local properties? I think I'm going to make `\g_xpg_langsetup_prop` local...

**Comment by Udi-Fogiel** [comment]
I implemented the idea in 11c1fe9. I tested the examples from #334 and #357 and it looks ok, please test when you have time.

**Comment by Udi-Fogiel** [comment]
@jspitz Is there anything left to do regarding this ticket?

**Comment by jspitz** [comment]
No, I think we are done.
